### DISCUSSION TITLE
How to add `babel-plugin-codegen` to build process

### DISCUSSION THREAD

**Discussion by jcao02** [nuxt/nuxt#9611] [Questions (Nuxt 2)]
Hello,

I'm trying to use [babel-plugin-codegen](https://github.com/kentcdodds/babel-plugin-codegen) in my project and I was wondering how can I make it run before the code is bundled for both client and server in all environments (dev and prod builds)

The use-case that triggered this question is that I want to get the middleware filename at build-time to be able to trace it:
```ts
// middlewares/my-middleware.ts
import { Context } from '@nuxt/types'
import { tracedRouterMiddleware } from '@/modules/observability'
import codegen from 'babel-plugin-codegen'

const filename = codegen`
    module.exports = require('path').basename()
`

export default tracedRouterMiddleware(async (ctx: Context) => {
  // ... Logic here
}, filename)

// @/modules/observability/index.ts
export function tracedRouterMiddleware(fn: Function, middlewareName: string) {
  // Tracing logic here
}
```

My first option was to get this at runtime on the server, but because everything's bundled, the filename is always `server.js`.

Any suggestions on how to proceed?

**Sidenote**: my team wants to leverage codegen for a lot of DX tooling/instrumentation and unlocking this would be a great improvement

**Comment by kissu** [comment]
You could call the script directly at the top of your `nuxt.config.js` file, if it's purely Node.js or with nothing related to Vue/Nuxt.

**Reply by jcao02** [reply]
Hey @kissu ! thanks for the quick reply. Can you elaborate more? You mean using [codegen Node's API](https://github.com/kentcdodds/babel-plugin-codegen#via-node-api)?

**Reply by kissu** [reply]
I was thinking more about this kind of thing: https://stackoverflow.com/a/67689890/8816585

**Reply by jcao02** [reply]
I understand. But I think that would modify the source code, which is not the desired outcome. The idea is to make the replacement at build time.

I tried adding the babel plugin to `nuxt.config.js` in `build.babel.plugins` array, but no luck. In the dev build, the files in the `.nuxt` folder point to my source code rather than a bundled version and in the prod build, the codegen is not replaced

**Reply by kissu** [reply]
You could make it conditional. Check if an environment variable is set to true and then run the code in `nuxt.config.js` or even do it in an external file and chain during the build time like `node my_build_script && yarn generate`.

**Reply by jcao02** [reply]
If I run babel previous nuxt build, that would modify the source code, wouldn't it? We'd need to target the babel output to some `dist` folder and then build nuxt on that same `dist` folder with the new files.

That's why I'm asking to chain it in the same build that nuxt uses, so we avoid that intermediate step

**Reply by kissu** [reply]
Why not `yarn generate && node my_build_script` so?

**Reply by jcao02** [reply]
Let me try that and I'll get back to you 👍🏼  Thanks for your help @kissu !
